cirq-core/cirq/ops/conditional_operation_test.py

```diff
@@ -0,0 +1,353 @@
+a: ════@╬═════^╬════
+        ║      ║
+b: ═════@══════^════
+      └──┘   └──┘
+""",
+        use_unicode_characters=True,
+    )
+
+
+def test_diagram_subcircuit():
+    q0, q1 = cirq.LineQubit.range(2)
+    circuit = cirq.Circuit(
+        cirq.CircuitOperation(
+            cirq.FrozenCircuit(
+                cirq.measure(q0, key='a'),
+                cirq.X(q1).with_conditions('a'),
+            )
+        )
+    )
+
+    cirq.testing.assert_has_diagram(
+        circuit,
+        """
+      Circuit_0x0000000000000000:
+      [ 0: ───M───────          ]
+0: ───[       ║                 ]───
+      [ 1: ───╫───X───          ]
+      [       ║   ║             ]
+      [ a: ═══@═══^═══          ]
+      │
+1: ───#2────────────────────────────
+""",
+        use_unicode_characters=True,
+    )
+
+
+def test_diagram_subcircuit_layered():
+    q0, q1 = cirq.LineQubit.range(2)
+    circuit = cirq.Circuit(
+        cirq.measure(q0, key='a'),
+        cirq.CircuitOperation(
+            cirq.FrozenCircuit(
+                cirq.measure(q0, key='a'),
+                cirq.X(q1).with_conditions('a'),
+            ),
+        ),
+        cirq.X(q1).with_conditions('a'),
+    )
+
+    cirq.testing.assert_has_diagram(
+        circuit,
+        """
+          Circuit_0x0000000000000000:
+          [ 0: ───M───────          ]
+0: ───M───[       ║                 ]───────
+      ║   [ 1: ───╫───X───          ]
+      ║   [       ║   ║             ]
+      ║   [ a: ═══@═══^═══          ]
+      ║   ║
+1: ───╫───#2────────────────────────────X───
+      ║   ║                             ║
+a: ═══@═══╩═════════════════════════════^═══
+""",
+        use_unicode_characters=True,
+    )
+
+
+def test_qasm():
+    q0, q1 = cirq.LineQubit.range(2)
+    circuit = cirq.Circuit(cirq.measure(q0, key='a'), cirq.X(q1).with_conditions('a'))
+    qasm = cirq.qasm(circuit)
+    assert (
+        qasm
+        == """// Generated from Cirq v0.14.0.dev
+
+OPENQASM 2.0;
+include "qelib1.inc";
+
+
+// Qubits: [0, 1]
+qreg q[2];
+creg m_a[1];
+
+
+measure q[0] -> m_a[0];
+if (m_a!=0) x q[1];
+"""
+    )
+
+
+@pytest.mark.parametrize('sim', ALL_SIMULATORS)
+def test_key_unset(sim):
+    q0, q1 = cirq.LineQubit.range(2)
+    circuit = cirq.Circuit(
+        cirq.measure(q0, key='a'),
+        cirq.X(q1).with_conditions('a'),
+        cirq.measure(q1, key='b'),
+    )
+    result = sim.run(circuit)
+    assert result.measurements['a'] == 0
+    assert result.measurements['b'] == 0
+
+
+@pytest.mark.parametrize('sim', ALL_SIMULATORS)
+def test_key_set(sim):
+    q0, q1 = cirq.LineQubit.range(2)
+    circuit = cirq.Circuit(
+        cirq.X(q0),
+        cirq.measure(q0, key='a'),
+        cirq.X(q1).with_conditions('a'),
+        cirq.measure(q1, key='b'),
+    )
+    result = sim.run(circuit)
+    assert result.measurements['a'] == 1
+    assert result.measurements['b'] == 1
+
+
+@pytest.mark.parametrize('sim', ALL_SIMULATORS)
+def test_subcircuit_key_unset(sim):
```

Comment:
Would tests of `CircuitOperation(...).with_conditions(...)` be interesting?

Reply:
Not really. A subop is a subop. But I can add one.

Reply:
Actually, I think that would just burn CPU unnecessarily. More useful would be the test that prohibits circuitoperations with measurements as suboperations. (If we decide to prohibit them).